File: app/src/main/java/com/travlog/android/apps/viewmodels/SignUpViewModel.java
package com.travlog.android.apps.viewmodels;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Patterns;

import com.travlog.android.apps.libs.ActivityViewModel;
import com.travlog.android.apps.libs.Environment;
import com.travlog.android.apps.libs.rx.Optional;
import com.travlog.android.apps.services.ApiClientType;
import com.travlog.android.apps.services.apirequests.XauthBody;
import com.travlog.android.apps.services.apiresponses.AccessTokenEnvelope;
import com.travlog.android.apps.services.apiresponses.Envelope;
import com.travlog.android.apps.ui.activities.SignUpActivity;
import com.travlog.android.apps.viewmodels.errors.SignUpViewModelErrors;
import com.travlog.android.apps.viewmodels.inputs.SignUpViewModelInputs;
import com.travlog.android.apps.viewmodels.outputs.SignUpViewModelOutputs;

import java.util.regex.Pattern;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.CompletableSubject;
import io.reactivex.subjects.PublishSubject;
import timber.log.Timber;

import static com.travlog.android.apps.libs.rx.transformers.Transformers.combineLatestPair;
import static com.travlog.android.apps.libs.rx.transformers.Transformers.pipeApiErrorsTo;
import static com.travlog.android.apps.libs.rx.transformers.Transformers.takeWhen;

public class SignUpViewModel extends ActivityViewModel<SignUpActivity>
        implements SignUpViewModelInputs, SignUpViewModelOutputs, SignUpViewModelErrors {

    private final ApiClientType apiClient;

    @SuppressLint("CheckResult")
    public SignUpViewModel(final @NonNull Environment environment) {
        super(environment);

        this.apiClient = environment.apiClient;

        final Observable<Boolean> isValidEmail = this.email.map(this::isValidEmail);

        final Observable<Boolean> isValidPassword = this.password.map(this::isValidPassword);

        Observable.merge(
                isValidEmail,
                isValidPassword
        )
                .compose(bindToLifecycle())
                .subscribe(setNextButtonEnabled);


        email.compose(combineLatestPair(password))
                .compose(takeWhen(signUpClick))
                .compose(bindToLifecycle())
                .switchMap(ep -> this.signUp(ep.first, ep.second))
                .compose(bindToLifecycle())
                .subscribe(envelope -> {
                    environment.currentUser.login(envelope.data.user, envelope.data.accessToken);
                    signUpSuccess.onComplete();
                });

        duplicatedError()
                .compose(bindToLifecycle())
                .subscribe(msg -> Timber.d("SignUpViewModel: duplicatedError: %s", msg));
    }

    private boolean isValidEmail(final @NonNull String email) {
        return Pattern.matches(Patterns.EMAIL_ADDRESS.pattern(), email);
    }

    private boolean isValidPassword(final @NonNull String password) {
        return !TextUtils.isEmpty(password);
    }

    private @NonNull
    Observable<AccessTokenEnvelope> signUp(final @NonNull String email,
                                           final @NonNull String password) {

        final XauthBody body = new XauthBody();
        body.email = email;
        body.password = password;

        return apiClient.signup(body)
                .compose(pipeApiErrorsTo(signUpError))
//                .compose(neverError())
                .toObservable();
    }

    private final PublishSubject<String> email = PublishSubject.create();
    private final PublishSubject<String> password = PublishSubject.create();
    private final PublishSubject<Optional> signUpClick = PublishSubject.create();
    private final PublishSubject<Envelope> signUpError = PublishSubject.create();

    private final BehaviorSubject<Boolean> setNextButtonEnabled = BehaviorSubject.create();
    private final CompletableSubject signUpSuccess = CompletableSubject.create();

    public final SignUpViewModelInputs inputs = this;
    public final SignUpViewModelOutputs outputs = this;
    public final SignUpViewModelErrors errors = this;

    @Override
    public void email(final @NonNull String email) {
        this.email.onNext(email);
    }

    @Override
    public void password(final @NonNull String password) {
        this.password.onNext(password);
    }

    @Override
    public void signUpClick() {
        this.signUpClick.onNext(new Optional<>(null));
    }

    @NonNull
    @Override
    public Observable<Boolean> setNextButtonEnabled() {
        return setNextButtonEnabled;
    }

    @NonNull
    @Override
    public Completable signUpSuccess() {
        return signUpSuccess;
    }

    @NonNull
    @Override
    public Observable<String> duplicatedError() {
        return signUpError.filter(Envelope::isDuplicatedUser)
                .map(errorEnvelope -> errorEnvelope.err.msg);
    }
}

File: app/src/main/java/com/travlog/android/apps/viewmodels/SignUpViewModel.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.viewmodels

import android.annotation.SuppressLint
import android.util.Pair
import android.util.Patterns
import com.travlog.android.apps.libs.ActivityViewModel
import com.travlog.android.apps.libs.Environment
import com.travlog.android.apps.libs.rx.Optional
import com.travlog.android.apps.libs.rx.transformers.Transformers.combineLatestPair
import com.travlog.android.apps.libs.rx.transformers.Transformers.neverError
import com.travlog.android.apps.libs.rx.transformers.Transformers.pipeApiErrorsTo
import com.travlog.android.apps.libs.rx.transformers.Transformers.takeWhen
import com.travlog.android.apps.services.ApiClientType
import com.travlog.android.apps.services.apirequests.SignUpBody
import com.travlog.android.apps.services.apiresponses.AccessTokenEnvelope
import com.travlog.android.apps.services.apiresponses.Envelope
import com.travlog.android.apps.ui.activities.SignUpActivity
import com.travlog.android.apps.viewmodels.errors.SignUpViewModelErrors
import com.travlog.android.apps.viewmodels.inputs.SignUpViewModelInputs
import com.travlog.android.apps.viewmodels.outputs.SignUpViewModelOutputs
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.CompletableSubject
import io.reactivex.subjects.PublishSubject
import timber.log.Timber
import java.util.regex.Pattern
import javax.inject.Inject

@SuppressLint("CheckResult")
class SignUpViewModel @Inject constructor(environment: Environment
) : ActivityViewModel<SignUpActivity>(environment), SignUpViewModelInputs, SignUpViewModelOutputs, SignUpViewModelErrors {

    private val apiClient: ApiClientType = environment.apiClient

    private val email = PublishSubject.create<String>()
    private val password = PublishSubject.create<String>()
    private val signUpClick = PublishSubject.create<Optional<*>>()
    private val signUpError = PublishSubject.create<Envelope>()

    private val setNextButtonEnabled = BehaviorSubject.create<Boolean>()
    private val signUpSuccess = CompletableSubject.create()

    val inputs: SignUpViewModelInputs = this
    val outputs: SignUpViewModelOutputs = this
    val errors: SignUpViewModelErrors = this

    init {
        val isValidEmail = this.email.map { this.isValidEmail(it) }

        val isValidPassword = this.password.map { this.isValidPassword(it) }

        Observable.merge(
                isValidEmail,
                isValidPassword
        )
                .compose(bindToLifecycle())
                .subscribe { setNextButtonEnabled.onNext(it) }


        val emailAndPassword: Observable<Pair<String, String>> = email.compose(combineLatestPair(password))

        emailAndPassword
                .compose<Pair<String, String>>(takeWhen(signUpClick))
                .switchMap {
                    this.signUp(it.first, it.second)
                            .doOnSubscribe {

                            }
                            .doAfterTerminate {

                            }
                }
                .compose(bindToLifecycle())
                .subscribe {
                    environment.currentUser.login(it.data.user, it.data.accessToken)
                    signUpSuccess.onComplete()
                }

        duplicatedError()
                .compose(bindToLifecycle())
                .subscribe { msg -> Timber.d("SignUpViewModel: duplicatedError: %s", msg) }
    }

    private fun isValidEmail(email: String): Boolean {
        return Pattern.matches(Patterns.EMAIL_ADDRESS.pattern(), email)
    }

    private fun isValidPassword(password: String): Boolean {
        return password.isNotEmpty()
    }

    private fun signUp(email: String, password: String): Observable<AccessTokenEnvelope> {

        val body = SignUpBody()
        body.email = email
        body.password = password

        return apiClient.signUp(body)
                .compose(pipeApiErrorsTo(signUpError))
                .compose(neverError())
                .toObservable()
    }

    override fun email(email: String) {
        this.email.onNext(email)
    }

    override fun password(password: String) {
        this.password.onNext(password)
    }

    override fun signUpClick() {
        this.signUpClick.onNext(Optional<Any>(null))
    }

    override fun setNextButtonEnabled(): Observable<Boolean> {
        return setNextButtonEnabled
    }

    override fun signUpSuccess(): Completable {
        return signUpSuccess
    }

    override fun duplicatedError(): Observable<String> {
        return signUpError.filter { it.isDuplicatedUser }.map { it.err.msg }
    }
}
